src/main/java/no/uib/inf112/core/map/GameMap.java

```diff
@@ -110,41 +133,86 @@ public TiledMapTileSets getMapTileSets() {
 
     @Override
     public void addEntity(@NotNull Entity entity) {
-        for (Entity knownRobot : getEntities()) {
-            if (entity.getX() == knownRobot.getX() && entity.getY() == knownRobot.getY()) {
-                throw new IllegalStateException("Cannot add an entity on top of another entity");
-            }
+        if (getTile(entityLayer, entity.getX(), entity.getY()) != null) {
+            throw new IllegalStateException("Cannot add an entity on top of another entity");
         }
-        entities.put(entity, null);
+        entities.put(new UVector2Int(entity.getX(), entity.getY()), entity);
+        entity.update(true);
     }
 
+    @Override
+    public boolean isOutsideBoard(int x, int y) {
+        return x < 0 || x >= getMapWidth() || y < 0 | y >= getMapHeight();
+    }
 
+    //TODO ISSUE #118 this should be tested
     @Override
-    @Nullable
-    public Entity getEntity(int x, int y) {
-        Vector2Int v = new Vector2Int(x, y);
-        for (Map.Entry<Entity, Vector2Int> entry : entities.entrySet()) {
-            if (v.equals(entry.getValue())) {
-                return entry.getKey();
-            }
+    public boolean removeEntity(@Nullable Entity entity) {
+        if (entity == null) {
+            return false;
+        }
+        UVector2Int pos = new UVector2Int(entity.getX(), entity.getY());
+        Entity e = entities.get(pos);
+
+        if (entity.equals(e)) {
+            entityLayer.setCell(entity.getX(), entity.getY(), null);
+            entities.remove(pos);
         }
-        return null;
+        return false;
     }
 
     @Override
-    public boolean removeEntity(Entity entity) {
-        return entities.remove(entity) == null;
+    public void addEntityLaser(@NotNull Tile laser) {
+        for (Tile knownLaser : getLaserEntities()) {
+
+            if (laser.getX() == knownLaser.getX() && laser.getY() == knownLaser.getY()) {
+                // Already a laser tile in this layer, se if we need to change it to a cross or just ignore it (same orientation)
+                if (laser.getTile() != knownLaser.getTile()) {
+                    removeEntityLaser(knownLaser);
+                    entityLasers.add(new LaserTile(new Vector2Int(knownLaser.getX(), knownLaser.getY()), TileGraphic.LASER_CROSS, Color.WHITE));
+                    entityLaserLayer.setCell(laser.getX(), laser.getY(), new TiledMapTileLayer.Cell().setTile(TileGraphic.LASER_CROSS.getTile()));
+                    return;
+                } else {
+                    /* We should probably have a count for when this happens, cleanup will try and remove non existing tile since we didn't add it.
```

Comment:
Issue?

Reply:
Currently it is not an issue as this doesn't cause any problems, but its important for the developer to be aware of this sideeffect